packages/material-ui/src/utils/useEventCallback.js

```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+
+const useEnhancedEffect = typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;
+
+/**
+ * https://github.com/facebook/react/issues/14099#issuecomment-440013892
+ *
+ * @param {function} fn
+ */
+export default function useEventCallback(fn) {
```

Comment:
@oliviertassinari do we know why we should be using `useLayoutEffect` instead of `useEffect` (https://reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback)? I guess it's for consistency and in case there's any handlers that fire straight after rendering?

Reply:
Yes, I think that you are on the right track. You would want the handler to be as up-to-date as possible.

Reply:
Answering because I introduced it and thought about it: I'm not certain how the input priority is currently handled but from what I understand how effects are currently scheduled the following happens:
1. react renders (stale callback)
2. react commits (during commit queued layout effects from `useLayoutEffect` are processed)
3. react processes queued effects from `useEffect`

I think 3 can happen after user input i.e. between 2 and 3 a user input can happen. If we would update the callback with `useEffect` the callback would be stale. 

So `useLayoutEffect` is simply the safer option. At least it's never wrong to use `useLayoutEffect` over `useEffect`. `useEffect` allows for performance optimization since it gives react time to schedule it accordingly. Since it's only an assignement performance considerations aren't really relevant.

Would be nice if someone could investigate the PR that added the example and find out why it uses `useEffect` instead of `useLayoutEffect` and considered race conditions with user inputs.

Reply:
> Would be nice if someone could investigate the PR that added the example and find out why it uses useEffect instead of useLayoutEffect and considered race conditions with user inputs.

Could you link this PR, So everybody has the context :)

Reply:
> the PR that added the example

Which PR is this? Do you mean when you first introduced it? But yeah, I agree with your analysis - that was my gut feeling as well (esp. about a simple assignment having no perf impact). It' just good to hear it from someone else too 😅

Reply:
Oh I don't know if there was one. Just that reactjs.org is in version control with a public github repo: https://github.com/reactjs/reactjs.org

I hoped somebody else would do the legwork while I can keep on talking intelligently about scheduling :smile: